### PR TITLE
Update type inference to work in presence of ConfigureAwait

### LINKED ISSUE #643: Type inference for GFU and friends doesn't infer method to return Task if ConfigureAwait is used
Ported from TFS WorkItem: <b>809537</b>

------------------------------

<p><b>Repro Steps:</b></p>
<P>1. Add the following inside Main() inside a new console app:</P><FONT size=2 face=Consolas><FONT size=2 face=Consolas>
<P></FONT></FONT><FONT color=#0000ff size=2 face=Consolas><FONT color=#0000ff size=2 face=Consolas><FONT color=#0000ff size=2 face=Consolas></FONT></FONT></FONT>&nbsp;</P>
<P><FONT color=#0000ff size=2 face=Consolas><FONT color=#0000ff size=2 face=Consolas><FONT color=#0000ff size=2 face=Consolas>bool</FONT></FONT></FONT><FONT size=2 face=Consolas><FONT size=2 face=Consolas> x = </FONT></FONT><FONT color=#0000ff size=2 face=Consolas><FONT color=#0000ff size=2 face=Consolas><FONT color=#0000ff size=2 face=Consolas>await</FONT></FONT></FONT><FONT size=2 face=Consolas><FONT size=2 face=Consolas> Foo().ConfigureAwait(</FONT></FONT><FONT color=#0000ff size=2 face=Consolas><FONT color=#0000ff size=2 face=Consolas><FONT color=#0000ff size=2 face=Consolas>false</FONT></FONT></FONT><FONT size=2 face=Consolas><FONT size=2 face=Consolas>);</P></FONT></FONT>
<P>&nbsp;</P>
<P>2. Try GFU on Foo()</P>
<P>&nbsp;</P>
<P><STRONG>Expected:</STRONG> it infers a return type of Task&lt;bool&gt;</P>
<P><STRONG>Actual:</STRONG> it infers a return type of object.</P>
<P>&nbsp;</P>
<P>NB that if we remove the configure await, we do infer Task&lt;bool&gt;. It looks like the inference service needs to be caught about ConfigureAwait specifically, or more generally if you're awaiting the result of a method, we might want to see if that method is defined or is an extension of Task or Task&lt;T&gt;. The latter algorithm would also allow for things like:</P>
<P>&nbsp;</P>
<P>bool x = await Foo().SafeContinueWith(false)</P>
<P>&nbsp;</P>
<P>to infer something more useful.</P>

------------------------------

<p><b>Revisions:</b></p>

1) Created By Jason Malinowski (10/24/2013 5:40:24 PM)

------------------------------

2) Edited By Dustin Campbell (8/12/2014 3:52:21 PM)

Diagnostics team -- it looks like this one might be more appropriate on your plate?

------------------------------

3) Edited By Dustin Campbell (8/12/2014 3:52:46 PM)

Eh.. maybe not.

------------------------------

4) Edited By Kevin Pilch-Bisson (2/16/2015 3:45:10 PM)

Moving to Diagnostics for GFU.

------------------------------

### PR DESCRIPTION
**Pull request by jmarolf**
Fixes #643

### CODE REVIEW

**Comment by jmarolf** [comment]
@dpoeschl @srivatsn @tmeschter @shyamnamboodiripad @heejaechang @mavasani @JohnHamby Could you take a look please?

**Comment by srivatsn** [comment]
:+1:

**Comment by JohnHamby** [comment]
Assuming that Parent properties cannot be null, I sign off.

File: src/EditorFeatures/CSharpTest/Diagnostics/GenerateMethod/GenerateMethodTests.cs [hunk 1]
```diff
@@ -2685,6 +2685,15 @@ public void TestGenerateMethodInDictionaryInitilizer3()
 @"using System ; using System . Collections . Generic ; class Program { static void Main ( string [ ] args ) { var x = new Dictionary < string , int > { [ ""Zero"" ] = i ( ) } ; } private static int i ( ) { throw new NotImplementedException ( ) ; } } ");
         }
 
+        [WorkItem(643, "https://github.com/dotnet/roslyn/issues/643")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public void TestGenerateMethodWithConfigureAwaitFalse()
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/Diagnostics/GenerateMethod/GenerateMethodTests.cs]: Can you test the type inferrer directly as well (via TypeInferrerTests)?

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs [hunk 2]
```diff
@@ -193,6 +193,13 @@ private IEnumerable<ITypeSymbol> InferTypesWorker(ExpressionSyntax expression)
                     parent = parent.Parent;
                 }
 
+                if (parent is MemberAccessExpressionSyntax &&
```
Comment (dpoeschl) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs]: IsParentKind handles the null concerns @JohnHamby brought up

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs [hunk 3]
```diff
@@ -193,6 +193,13 @@ private IEnumerable<ITypeSymbol> InferTypesWorker(ExpressionSyntax expression)
...
+                    parent.Parent is InvocationExpressionSyntax &&
```
Comment (dpoeschl) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs]: This is a bit too particular. I think you'll miss cases like ```bool x = await [|M|]..ContinueWith(a => { return true; }).ContinueWith(a => { return false; });```

**Pull request merged**